### DISCUSSION TITLE
Automate Install Microsoft Build of OpenJDK with PowerShell

### DISCUSSION THREAD

**Discussion by MSIH** [microsoft/openjdk#278] [Q&A] [answered]
Due to security requirements, we need to upgrade java 8, 11, and 17, every month on 100s of servers running Windows 2012, 2016, and 2019. We are trying to automate the process using PowerShell.

- Best : a single PowerShell command to download, then install or upgrade java, and then setup JAVA_HOME, and then Add to System Path.
- Second Best: vanity URL for that downloads the latest msi file for java 8, 11, and 17

**Comment by brunoborges** [comment]
Hi @MSIH,

You can use the following command in PowerShell to download a JDK:

```powershell
Invoke-WebRequest -Uri https://aka.ms/download-jdk/microsoft-jdk-17-windows-x64.zip -OutFile jdk-17.zip -UseBasicParsing
Expand-Archive -Force jdk-17.zip -DestinationPath c:\
Remove-Item jdk-17.zip -Force
```

All vanity URLs are published here: https://docs.microsoft.com/en-us/java/openjdk/download-major-urls

**Reply by MSIH** [reply]
Thanks for the example. We are doing something like this already. The problem with the ZIP files is when you expand it you get a folder that includes the Java version number.  Which makes setting the Path and JAVA_HOME more difficult. 

The MSI file takes care of both setting and removing these environment variables.

Expand microsoft-jdk-11.0.15-windows-x64.zip and you get \jdk-11.0.15+10

**Reply by brunoborges** [reply]
So, you saying it would be good to have a URL download for the MSI package, without the minor version?

The other approach is to simply rename the folder extracted from the ZIP to something you already have set in your `PATH`.

Renaming an object without its full name (possibly using regex), is possible with Powershell.

We may be able to add vanity URLs for MSI packages in July.

**Reply by ecki** [reply]
If you plan to roll out sich a thing to hundreds machines you should locally archive and test it anyway, which makes a re-packaging step quite easy. So yes it would be good if Powershell could skip parent folder in archives, but for your process that won’t fly. Not to mention checksums, signatures, pre-config and hardening.

Since in place updates are not so atomic, expanding to a new dir and setting environment variables (in Powershell) sounds however like a good approach (second best would be a symlink - much better than replacing files of a running application.

**Reply by rbeesley** [reply]
I'd also recommend utilizing [winget](https://docs.microsoft.com/en-us/windows/package-manager/winget/) to manage this.

```PowerShell
> winget install Microsoft.OpenJDK.17
```

An advantage of using winget is that you can easily update to later version numbers by simply running:

```PowerShell
> winget upgrade Microsoft.OpenJDK.17
```

Both of these should probably be executed from an elevated PowerShell instance or you will receive UAC prompts, but winget seems like a very reliable way to manage things.

**Comment by MSIH** [comment, marked as answer]
I am not sure I understand the logic of not publishing the msi files and all the LTS versions of java (8,11,17)  on the vanity URL page. They already have the files and probably the automation. The URL on that page are only for Windows and windows uses MSI to install programs and tools.

**Reply by brunoborges** [reply]
Because you can use WinGet for that.

See https://docs.microsoft.com/en-us/java/openjdk/install#install-with-the-windows-package-manager

**Reply by MSIH** [reply]
Thanks for the suggestion. I looked into using that tool because nearly exactly what I was looking for but I do not think WinGet is supported on Windows Server yet. "[The winget command line tool is only supported on Windows 10 1709 (build 16299) or later at this time."](https://github.com/microsoft/winget-cli/issues/144#issuecomment-926158387)

"[We aren't targeting Windows Server 2019 as a supported SKU for the Windows Package Manager. There are several challenges with dependencies like the Store for updates, App Execution Aliases. You may be able to overcome the hurdles and get the Windows Package Manager installed, but it is not supported in this environment."](https://github.com/microsoft/winget-cli/issues/144#issuecomment-926158387)

**Reply by brunoborges** [reply]
Got it.

I'll work with the team to produce vanity URLs for the other installers then.

Hopefully we can get this done by Friday.

I'll create an issue soon and link here.

**Reply by brunoborges** [reply]
For the minor version in the folder, I found a nice way to rename it:

https://github.com/sirredbeard/nim-windows-container/blob/main/Dockerfile#L23

**Reply by MSIH** [reply]
Please provide a status update. I still do not see a vanity URL to download the latest version of java 11 msi file.

**Reply by karianna** [reply]
@d3r3kk - Can you post the aka.ms links here or where @MSIH can look them up?

**Reply by d3r3kk** [reply]
Will have them here by EOD.

**Reply by d3r3kk** [reply]
# Windows MSI Major Version Vanity URLs v1.0

## ✅ Windows 11 MSI x64
- https://aka.ms/download-jdk/microsoft-jdk-11-windows-x64.msi
- https://aka.ms/download-jdk/microsoft-jdk-11-windows-x64.msi.sha256sum.txt

## ✅ Windows 11 MSI AArch64
- https://aka.ms/download-jdk/microsoft-jdk-11-windows-aarch64.msi
- https://aka.ms/download-jdk/microsoft-jdk-11-windows-aarch64.msi.sha256sum.txt

## ✅ Windows 17 MSI x64
- https://aka.ms/download-jdk/microsoft-jdk-17-windows-x64.msi
- https://aka.ms/download-jdk/microsoft-jdk-17-windows-x64.msi.sha256sum.txt

## ✅ Windows 17 MSI AArch64
- https://aka.ms/download-jdk/microsoft-jdk-17-windows-aarch64.msi 
- https://aka.ms/download-jdk/microsoft-jdk-17-windows-aarch64.msi.sha256sum.txt

...we will allow our hero @brunoborges to publicize these in a more fashionable manner upon his return.

**Reply by MSIH** [reply]
Thank you

**Reply by MSIH** [reply]
When are you going to up this page: https://docs.microsoft.com/en-us/java/openjdk/download-major-urls

**Reply by d3r3kk** [reply]
Thanks for the reply (and admittedly, the reminder that I need to follow through and add this to the public page!).

However, we are in the midst of a re-spin of the 17 and 11 binaries so all release activity will be completed (re-completed?) in the next week or so, and at that time I will update this page. 

For the interim, please use the links above in this issue.

**Reply by d3r3kk** [reply]
This is complete now (updated it some time ago, just after our respin release).
